modules/ngx_debug_conn/ngx_http_debug_conn_module.c

```diff
@@ -0,0 +1,277 @@
+                                " handler: r:%p w:%p\n"   \
+                                "requests: %ui\n"         \
+                                "poolsize: %z\n"
+
+#define NGX_REQ_ENTRY_SIZE      (sizeof(NGX_REQ_ENTRY_FORMAT) - 1 + \
+                                 sizeof("https") - 1 + 32 /* host */ + 64 /* uri */ + \
+                                 NGX_TIME_T_LEN + NGX_PTR_SIZE * 2 * 2 + NGX_SIZE_T_LEN)
+#define NGX_REQ_ENTRY_FORMAT    "********* request ******\n"\
+                                "     uri: %s://%V%V\n"   \
+                                " handler: r:%p w:%p\n"   \
+                                "startsec: %T\n"          \
+                                "poolsize: %z\n"
+
+    n = ngx_cycle->connection_n - ngx_cycle->free_connection_n;
+
+    size = NGX_CONN_TITLE_SIZE + n * (NGX_CONN_ENTRY_SIZE + NGX_REQ_ENTRY_SIZE);
+    p = ngx_palloc(pool, size);
+    if (p == NULL) {
+        return NGX_ERROR;
+    }
+
+    b->pos = p;
+
+    p = ngx_sprintf(p, NGX_CONN_TITLE_FORMAT, ngx_pid, n);
+
+    /* lines of entry */
+
+    k = 0;
+
+    for (i = 0; i < ngx_cycle->connection_n; i++) {
+        c = &ngx_cycle->connections[i];
+
+        if (c->fd <= 0) {
+            continue;
+        }
+
+        k++;
+
+        if (n == 0) {
+            break;
+        }
+        n--;
+
+        /* addr_text */
+
+        if (c->addr_text.data != NULL) {
+            addr.data = c->addr_text.data;
+            addr.len = ngx_min(c->addr_text.len, NGX_SOCKADDR_STRLEN);
+
+        } else if (c->listening && c->listening->addr_text.data != NULL) {
+            addr.data = c->listening->addr_text.data;
+            addr.len = ngx_min(c->listening->addr_text.len, NGX_SOCKADDR_STRLEN);
+
+        } else {
+            ngx_str_set(&addr, "(null)");
+        }
+
+        /* action */
+
+        if (c->log->action != NULL) {
+            action.data = (u_char *) c->log->action;
+            action.len = ngx_min(ngx_strlen(c->log->action), 32);
+
+#if (NGX_SSL)
+        } else if (c->ssl) {
+
+            ngx_str_set(&action, "(null: ssl)");
+#endif
+
+        } else if (c->listening && c->listening->connection == c) {
+            ngx_str_set(&action, "(null: listening)");
+
+        } else if (c->data == NULL) {
+            ngx_str_set(&action, "(null: channel)");
+
+        } else {
+            ngx_str_set(&action, "(null)");
+        }
+
+        /* entry format of connection */
+
+        p = ngx_snprintf(p, NGX_CONN_ENTRY_SIZE, NGX_CONN_ENTRY_FORMAT,
+                         k, i,
+                         c->fd,
+                         &addr,
+                         c->sent,
+                         &action,
+                         c->read->handler, c->write->handler,
+                         c->requests,
+                         c->pool ? ngx_pool_size(c->pool) : (size_t) 0);
+
+        /* c->data: http request */
+
+        if (c->data != NULL) {
+            r = (ngx_http_request_t *) c->data;
```

Comment:
It seems that the `c->data` could be `ngx_http_request_t`, `ngx_http_v2_connection_t`, `ngx_stream_session_t` or other stuff filled by the Nginx module. We could not safely assume it must be `ngx_http_request_t`.

I comment here because I am looking up a way to get the `ngx_http_request_t` via a given `ngx_connection_t`. If you find one, please let me know.

Reply:
Yes, it's known issue. 

> I comment here because I am looking up a way to get the ngx_http_request_t via a given ngx_connection_t. If you find one, please let me know.

It seems that there is no good way to determine c->data type.
Maybe we add `ngx_connection_t:data_type` field, which is not good way.